tests/unit/s2n_certificate_test.c

```diff
@@ -0,0 +1,84 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+#include "s2n_test.h"
+#include "testlib/s2n_testlib.h"
+
+#include <s2n.h>
+#include "utils/s2n_safety.h"
+
+#define S2N_DEFAULT_TEST_CERT_CHAIN_LENGTH 3
+
+int main(int argc, char **argv)
+{
+    BEGIN_TEST();
+
+    struct s2n_cert_chain_and_key *chain_and_key = NULL;
+    EXPECT_SUCCESS(
+        s2n_test_cert_chain_and_key_new(&chain_and_key, S2N_DEFAULT_TEST_CERT_CHAIN, S2N_DEFAULT_TEST_PRIVATE_KEY));
+
+    /* Test s2n_get_cert_chain_length */ 
+    {
+        uint32_t length = 0;
+        EXPECT_FAILURE_WITH_ERRNO(s2n_get_cert_chain_length(NULL, &length), S2N_ERR_NULL);
+        EXPECT_SUCCESS(s2n_get_cert_chain_length(chain_and_key, &length));
+        EXPECT_EQUAL(length, S2N_DEFAULT_TEST_CERT_CHAIN_LENGTH);
+    }
+
+    /* Test s2n_get_cert_from_cert_chain */
+    {
+        struct s2n_cert *out_cert = NULL;
+        uint32_t cert_idx = 0;
+
+        /* Safety checks */
+        {
+            EXPECT_FAILURE_WITH_ERRNO(s2n_get_cert_from_cert_chain(NULL, &out_cert, cert_idx), S2N_ERR_NULL);
+            EXPECT_FAILURE_WITH_ERRNO(s2n_get_cert_from_cert_chain(chain_and_key, NULL, cert_idx), S2N_ERR_NULL);
+            /* The valid range of cert_idx is 0 to cert_chain_length - 1 */  
+            cert_idx = S2N_DEFAULT_TEST_CERT_CHAIN_LENGTH; 
+            EXPECT_FAILURE_WITH_ERRNO(s2n_get_cert_from_cert_chain(chain_and_key, &out_cert, cert_idx), S2N_ERR_SAFETY);
+        }
+
+        struct s2n_cert *cur_cert = chain_and_key->cert_chain->head;
+
+        for (size_t i = 0; i < S2N_DEFAULT_TEST_CERT_CHAIN_LENGTH; i++)
+        {
+            EXPECT_SUCCESS(s2n_get_cert_from_cert_chain(chain_and_key, &out_cert, i));
+            EXPECT_NOT_NULL(cur_cert);
+            EXPECT_EQUAL(out_cert, cur_cert);
+            cur_cert = cur_cert->next;
+        }
+    }
+
+    /* Test s2n_get_cert_der */ 
+    {
```

Comment:
Since we have [an existing API that gets the der version of the client certificate](https://github.com/aws/s2n-tls/blob/main/api/s2n.h#L406), it might be worth testing that these APIs are equivalent to that API. That's behavior that should be guaranteed by a test.

The lack of an API to get the peer's chain might be an issue though, so you might need to just make a note that that test should be included once the "get peer's cert chain" API exists.

Reply:
The existing [s2n_connection_get_client_cert_chain] (https://github.com/aws/s2n-tls/blob/main/api/s2n.h#L406) API returns the DER of the entire cert chain and not a single cert in the cert chain. This API obtains the DER of a single cert in the cert chain. Therefore, they do not match and cannot be tested here.